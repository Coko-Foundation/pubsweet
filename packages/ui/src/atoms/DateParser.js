import moment from 'moment'
import propTypes from 'prop-types'
import { compose, withProps, withHandlers, setDisplayName } from 'recompose'

const getDuration = timestamp => {
  const today = moment()
  const stamp = moment(timestamp)
  return moment.duration(today.diff(stamp))
}

const D = ({ children, timestamp, timeAgo }) => children(timestamp, timeAgo)

const DateParser = compose(
  setDisplayName('DateParser'),
  withHandlers({
    renderTimestamp: ({
      timestamp,
      dateFormat = 'DD.MM.YYYY',
      humanizeThreshold = 0,
    }) => () => {
      if (!timestamp) return ''
      const duration = getDuration(timestamp)

      if (duration.asDays() < humanizeThreshold) {
        return `${duration.humanize()} ago`
      }
      return moment(timestamp).format(dateFormat)
    },
    renderTimeAgo: ({ timestamp }) => () => {
      if (!timestamp) return ''
      const duration = getDuration(timestamp)
      return duration.humanize()
    },
  }),
  withProps(({ renderTimestamp, renderTimeAgo }) => ({
    timeAgo: renderTimeAgo(),
    timestamp: renderTimestamp(),
  })),
)(D)

export default DateParser

DateParser.propTypes = {
  /** The date string. Can be any date parsable by momentjs. */
  timestamp: propTypes.oneOfType([propTypes.string, propTypes.number, Date])
    .isRequired,
  /** Format of the rendered date. */
  dateFormat: propTypes.string,
  /** Humanize duration threshold */
  humanizeThreshold: propTypes.number,
}
