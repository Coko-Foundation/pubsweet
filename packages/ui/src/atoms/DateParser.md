Component that renders a date string.

```js
<DateParser timestamp="2018-05-18T10:38:12.063Z">
  {(timestamp, daysAgo) => (
    <span>
      {timestamp} - {daysAgo}
    </span>
  )}
</DateParser>
```
