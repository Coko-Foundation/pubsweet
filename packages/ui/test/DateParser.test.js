import React from 'react'
import { mount, shallow } from 'enzyme'

import DateParser from '../src/atoms/DateParser'

describe('DateParser', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <DateParser timestamp="2018-05-18T10:38:12.063Z">
        {t => <span>{t}</span>}
      </DateParser>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('parses the date correctly', () => {
    let stamp
    mount(
      <DateParser timestamp="2018-05-18T10:38:12.063Z">
        {t => {
          stamp = t
          return <span>{t}</span>
        }}
      </DateParser>,
    )
    expect(stamp).toEqual('18.05.2018')
  })

  it('formats date correctly', () => {
    let stamp
    mount(
      <DateParser dateFormat="YYYY-DD-MM" timestamp="2018-05-18T10:38:12.063Z">
        {t => {
          stamp = t
          return <span>{t}</span>
        }}
      </DateParser>,
    )
    expect(stamp).toEqual('2018-18-05')
  })

  it('renders the right children', () => {
    const wrapper = mount(
      <DateParser dateFormat="YYYY-MM-DD" timestamp="2018-05-18T10:38:12.063Z">
        {t => <div>{t}</div>}
      </DateParser>,
    )
    expect(wrapper.find('div')).toHaveLength(1)
    expect(wrapper.html()).toMatch('2018-05-18')
  })

  it('renders with humanized threshold - 2 days ago', () => {
    const today = new Date()
    const fewDaysAgo = new Date().setDate(today.getDate() - 2)
    const wrapper = mount(
      <DateParser humanizeThreshold={3} timestamp={fewDaysAgo}>
        {t => <div>{t}</div>}
      </DateParser>,
    )
    expect(wrapper.find('div')).toHaveLength(1)
    expect(wrapper.html()).toMatch('<div>2 days ago</div>')
  })

  it('renders correct time ago', () => {
    const today = new Date()
    const wrapper = mount(
      <DateParser humanizeThreshold={3} timestamp={today}>
        {(t, ago) => <div>{ago} ago</div>}
      </DateParser>,
    )
    expect(wrapper.find('div')).toHaveLength(1)
    expect(wrapper.html()).toMatch('<div>a few seconds ago</div>')
  })
})
