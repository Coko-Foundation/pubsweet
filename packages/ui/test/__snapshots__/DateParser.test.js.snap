// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateParser renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DateParser
    timestamp="2018-05-18T10:38:12.063Z"
  >
    [Function]
  </DateParser>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
      "renderTimeAgo": [Function],
      "renderTimestamp": [Function],
      "timestamp": "2018-05-18T10:38:12.063Z",
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
        "renderTimeAgo": [Function],
        "renderTimestamp": [Function],
        "timestamp": "2018-05-18T10:38:12.063Z",
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
